ci_scripts/generate_news.py

```diff
@@ -1,64 +1,80 @@
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-# --------------------------------------------------------------------------
-"""Handles usage of towncrier for automated changelog generation
-
-a) abstracts configuration and execution for this project
-b) prepares for usage in CI environment to stop from generating
-   entries for every minor patch version
-
-However, when deploying to production, we always want a changelog
 """
+Handles usage of towncrier for automated changelog generation
+Handles usage of pyautoversion for versioning
+"""
+
 import os
 import subprocess
+import logging
+import enum
+import argparse
+from auto_version import auto_version_tool
+from utils.definitions import NEWS_DIR, PROJECT_ROOT, PROJECT_CONFIG, \
+    LOGGER_FORMAT
 
-from distutils.version import LooseVersion
+logger = logging.getLogger(__name__)
 
-PROJECT_ROOT = os.path.dirname(os.path.dirname(__file__))
-news_dir = os.path.join(PROJECT_ROOT, 'docs', 'news')
-metafile = os.path.join(news_dir, 'last_built.meta')
 
+class CommitType(enum.Enum):
+    DEVELOPMENT = 1
+    BETA = 2
+    RELEASE = 3
 
-def main():
-    """Writes out newsfile if significant version bump"""
-    last_known = '0'
-    if os.path.isfile(metafile):
-        with open(metafile) as fh:
-            last_known = fh.read()
-
-    import mbed_cloud
-    current = mbed_cloud.__version__
-
-    # how significant a change in version scheme should trigger a new changelog entry
-    # (api major, api minor, sdk major, sdk minor, sdk patch)
-    sigfigs = 4
-    current_version = LooseVersion(current).version
-    last_known_version = LooseVersion(last_known).version
-
-    should_towncrier = current_version[:sigfigs] != last_known_version[:sigfigs]
-
-    print('%s -- %s :: current vs previous changelog build' % (current, last_known))
-    if should_towncrier:
-        print('%s >> %s :: running changelog build' % (current, last_known))
+
+def version_project(commit_type: CommitType):
+    """
+    Versions the project
+    :param commit_type: states what is the type of the commit
+    :type: CommitType
+    """
+
+    use_news_files = commit_type in [CommitType.BETA, CommitType.RELEASE]
+
+    new_version = _calculate_version(commit_type, use_news_files)
+    _generate_changelog(new_version, use_news_files)
+
+
+def _calculate_version(commit_type: CommitType, use_news_files: bool):
+    BUMP_TYPES = {CommitType.DEVELOPMENT: "build",
+                  CommitType.BETA: "prerelease"}
+    old, new_version, updates = auto_version_tool.main(
+        release=commit_type == CommitType.RELEASE,
+        enable_file_triggers=True if use_news_files else None,
+        bump=BUMP_TYPES.get(commit_type),
```

Comment:
From a code style I would prefer to see parameters calculated before they are passed into the function.